src/nginx.c

```diff
@@ -234,35 +272,160 @@ static int nginx_read(void) {
+
+  DEBUG("nginx plugin: cleaning data for %s (%s).",
+        inst_config->url, inst_config->inst_name);
+
+  if (inst_config->curl != NULL)
+    curl_easy_cleanup(inst_config->curl);
+
+  sfree(cleanup_data);
+} /* void nginx_inst_cleanup */
+
+
+static int config_inst(char* inst_name, oconfig_item_t *inst_config_item) {
+  nginx_inst_config_t *inst_config;
+
+  oconfig_item_t *config_parameter;
+  char *inst_config_key;
+  char *inst_config_value;
+  int status;
+
+  inst_config = calloc(1, sizeof(*inst_config));
+
+  if (inst_config == NULL) {
+    ERROR("nginx plugin: calloc failed.");
+    return -1;
+  }
+
+  inst_config->host_name = NULL;
+  inst_config->inst_name = (inst_name==NULL) ? NULL : strdup(inst_name);
+
+  for (int i = 0; i < inst_config_item->children_num; i++) {
+    config_parameter = inst_config_item->children + i;
+    inst_config_key = config_parameter->key;
+    inst_config_value = config_parameter->values[0].value.string;
+    status =
+        config_inst_parameter(inst_config, inst_config_key, inst_config_value);
+    if (status != 0) {
+      ERROR("nginx plugin: config parsing failed.");
+      sfree(inst_config);
+      return -1;
+    }
+  }
+
+  status = init_curl(inst_config);
+  if (status != 0) {
+    ERROR("nginx plugin: curl initialization failed.");
+    sfree(inst_config);
+    return -1;
+  }
+
+  user_data_t ud;
+  char cb_name[DATA_MAX_NAME_LEN];
+
+  if (inst_config->inst_name == NO_INSTNAME) {
+    DEBUG("nginx plugin: Registering new read callback for single instance %s",
+          inst_config->url);
+    snprintf(cb_name, sizeof(cb_name), "%s", "nginx");
+  } else {
+    DEBUG("nginx plugin: Registering new read callback for instance %s (%s).",
+          inst_config->url, inst_config->inst_name);
+    snprintf(cb_name, sizeof(cb_name), "nginx.%s",
+             inst_config->inst_name);
+  }   
+
+  memset(&ud, 0, sizeof(ud));
+
+  ud.data = inst_config;
+  ud.free_func = nginx_inst_cleanup;
+
+  plugin_register_complex_read(NULL, cb_name, nginx_read, 0, &ud);
+
+  return 0;
+} /* config_inst */
+
+static int config_multiinstance(oconfig_item_t *ci) {
+
+  for (int i = 0; i < ci->children_num; i++) {
+    oconfig_item_t *child = ci->children + i;
+
+    if (strcasecmp("Instance", child->key) == 0) {
+
+      if ((child->values_num != 1) ||
+          (child->values[0].type != OCONFIG_TYPE_STRING)) {
+        WARNING("nginx plugin: The Instance block "
+                "needs exactly one string argument.");
+      } else {
+        if (config_inst(child->values[0].value.string, child) != 0)
+          return -1;
+      }
+
+    } else {
+      WARNING("nginx plugin: Ignoring unknown config option `%s'.  "
+              "expected 'Instance' block.", child->key);
+    }
+  }
+  return 0;
+} /* int config_multiinstance */
+
+static int config_singleinstance(oconfig_item_t *ci) {
+  return config_inst(NO_INSTNAME, ci);
+} /* int config_singleinstance */
+
+static int config_plugin(oconfig_item_t *ci) {
```

Comment:
Implementation which exists in `memcached` plugin is much more clear, do not introduces code duplication and excessive functions.